go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+func (bi *blameInfo) TimestampTime() time.Time {
+	return time.Unix(int64(bi.Timestamp/1000), 0)
+}
+
+// TimestampString returns a string representing the blameInfo timestamp
+func (bi *blameInfo) TimestampString() string {
+	return bi.TimestampTime().Format(time.UnixDate)
+}
+
+// A blame graph is a map of primary key hashes to blameInfo structs
+type blameGraph map[hash.Hash]blameInfo
+
+// Blame implements the `dolt blame` command. Blame annotates each row in the given table with information
+// from the revision which last modified the row, optionally starting from a given revision.
+//
+// Blame is computed as follows:
+//
+// First, a blame graph is initialized with one node for every row in the table at the given commit (defaulting
+// to HEAD of the currently checked-out branch).
+//
+// Starting from the given commit, walk backwards through the commit graph (currently by following each commit's
+// first parent, though this may change in the future).
+//
+// For each adjacent pair of commits `old` and `new`, check each remaining unblamed node to see if the row it represents
+// changed between the commits. If so, mark it with `new` as the blame origin and continue to the next node without blame.
+//
+// When all nodes have blame information, stop iterating through commits and print the blame graph.
+func Blame(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := argparser.NewArgParser()
+	help, usage := cli.HelpAndUsagePrinters(commandStr, blameShortDesc, blameLongDesc, blameSynopsis, ap)
+	apr := cli.ParseArgs(ap, args, help)
+
+	if apr.NArg() < 2 {
+		usage()
+		return 1
+	}
+
+	cs, err := parseCommitSpec(dEnv, apr)
+	if err != nil {
+		cli.PrintErr(err)
+		return 1
+	}
+
+	tableName := apr.Arg(1)
+
+	if err := runBlame(ctx, dEnv, cs, tableName); err != nil {
+		cli.PrintErr(err)
+		return 1
+	}
+
+	return 0
+}
+
+func runBlame(ctx context.Context, dEnv *env.DoltEnv, cs *doltdb.CommitSpec, tableName string) error {
+	commit, err := dEnv.DoltDB.Resolve(ctx, cs)
+	if err != nil {
+		return err
+	}
+
+	blameGraph, err := blameGraphFromCommit(ctx, dEnv, commit, tableName)
+	if err != nil {
+		return err
+	}
+
+	pkColNames, err := pkColNamesFromCommit(ctx, commit, tableName)
+	if err != nil {
+		return err
+	}
+
+	cli.Println(blameGraph.String(ctx, pkColNames))
+	return nil
+}
+
+func blameGraphFromCommit(ctx context.Context, dEnv *env.DoltEnv, commit *doltdb.Commit, tableName string) (*blameGraph, error) {
+	// get the commits sorted from newest to oldest ending with `commit`
+	commits, err := actions.TimeSortedCommits(ctx, dEnv.DoltDB, commit, -1)
+	if err != nil {
+		return nil, err
+	}
+
+	rows, err := rowsFromCommit(ctx, commit, tableName)
+	if err != nil {
+		return nil, err
+	}
+
+	blameGraph, err := blameGraphFromRows(ctx, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range commits {
+		// get the first parent
+		parent, err := dEnv.DoltDB.ResolveParent(ctx, c, 0)
```

Comment:
Would be great to package this up with the values in `commits` themselves since we're doing this same work every row * every commit.

Reply:
Boom: https://github.com/liquidata-inc/dolt/pull/157/commits/0665eb6fa9aff9442d12f20c33532893ee7e16f7